### PR TITLE
Add common package and use Rollup to bundle lib packages

### PR DESCRIPTION
**Pull request by vojtechszocs**
Fixes [HAC-383](https://issues.redhat.com/browse/HAC-383)

- add `common` package - code, types and configs used by other packages
- rename `lib-shared` package to `lib-utils` - Kubernetes and React utilities
- library packages now use [Rollup](https://github.com/rollup/rollup) to bundle their code and types (inspired by [this article](https://dev.to/sidv93/component-library-setup-with-react-typescript-and-rollup-onj))
- use [Rollup analyzer plugin](https://github.com/doesdev/rollup-plugin-analyzer) to show the generated bundle summary
- use [Rollup `d.ts` plugin](https://github.com/Swatinem/rollup-plugin-dts) to generate a single `d.ts` type declaration file
- runtime dependencies of library packages are declared exclusively via `peerDependencies`
- replace [Joi](https://github.com/sideway/joi) with [Yup](https://github.com/jquense/yup) due to Rollup reporting [circular dependencies originating from Joi](https://github.com/sideway/joi/issues/2517)
- allow `devDependencies` to be used in private packages ([import/no-extraneous-dependencies](https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/no-extraneous-dependencies.md))
- add `clean.sh` script to recursively prune `node_modules` and `dist` directories

Note 1: `"typescript": "~4.4.4"` is due to compatibility with currently used ESLint and its plugins.

Note 2: during `yarn install` there's a warning for an optional `fsevents` dependency (MacOS specific)
```
➤ YN0076: │ fsevents@patch:fsevents@npm%3A2.3.2#~builtin<compat/fsevents>::version=2.3.2&hash=18f3a7 The linux-x64 architecture is incompatible with this module, link skipped.
```
which comes from Rollup
```
$ yarn why fsevents
└─ rollup@npm:2.61.1
   └─ fsevents@patch:fsevents@npm%3A2.3.2#~builtin<compat/fsevents>::version=2.3.2&hash=18f3a7 (via patch:fsevents@~2.3.2#~builtin<compat/fsevents>)
```
AFAIK there's no way to suppress this warning, it's safe to ignore it.

### CODE REVIEW

File: packages/lib-runtime/rollup.config.js [hunk 1]
```diff
@@ -0,0 +1,8 @@
+// eslint-disable-next-line import/no-relative-packages
+import { tsLibConfig, dtsLibConfig } from '../common/rollup-configs';
```
Comment (vojtechszocs) [inline packages/lib-runtime/rollup.config.js:2]: The reason we're suppressing ESLint `import/no-relative-packages` rule here is due to default [Rollup config file transpile behavior](https://rollupjs.org/guide/en/#using-untranspiled-config-files).

We can revisit and fix this later on.

**Comment by andrewballantyne** [comment]
/cc @karelhala 

@florkbr @karelhala please offer any additional feedback. Seems you guys had a good chat back during the initial investigation, feel free to cap it off with a review 🙂

**Review by karelhala — APPROVED** [review]
looking good! The name of the package company is bit strange, is the `@monorepo` free? It might be better to create new org on npm with something like `@dynamic-plugin`.

File: packages/common/rollup-configs.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
+import { execSync } from 'child_process';
+/* eslint-disable import/no-extraneous-dependencies */
+import nodeResolve from '@rollup/plugin-node-resolve';
+import typescript from '@rollup/plugin-typescript';
+import analyzer from 'rollup-plugin-analyzer';
+import dts from 'rollup-plugin-dts';
+/* eslint-enable import/no-extraneous-dependencies */
```
[the remark this answers is not in the document]
Comment (vojtechszocs) [reply packages/common/rollup-configs.js:7]: Agreed, will do in a follow-up.

File: packages/common/rollup-configs.js [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  return (source) => externalModuleTests.some((regexp) => regexp.test(source));
+};
+
+const getBanner = (pkg) => `/*
+  OpenShift Dynamic Plugin SDK
+  ${pkg.repository.url.replace(/\.git$/, '')}
+
+  ${pkg.name} version ${pkg.version}
+  ${new Date().toLocaleString('en-US', { dateStyle: 'long', timeStyle: 'long' })}
+  commit ${execSync('git rev-parse HEAD').toString().trim()}
```
[the remark this answers is not in the document]
Comment (vojtechszocs) [reply packages/common/rollup-configs.js:26]: The build & publish of plugin SDK packages can be done manually, on-demand and per-package.

For example, we expect the `@openshift/dynamic-plugin-sdk-utils` package to change & get published more frequently than the `@openshift/dynamic-plugin-sdk-runtime` package.

What I meant in my previous comment is recommending to follow the pattern
- make changes to the given package
- bump version in package manifest
- have above changes merged into `main` branch (I don't think we need separate release branches)
- build & publish from `main` branch (manually for now, can be automated later via post-merge CI hook)

File: packages/common/tsconfig-bases/all-targets.json [hunk 4]
```diff
@@ -3,8 +3,6 @@
   "compilerOptions": {
     "noEmitOnError": true,
     "declaration": true,
-    "declarationMap": true,
-    "sourceMap": true,
```
[the remark this answers is not in the document]
Comment (vojtechszocs) [reply packages/common/tsconfig-bases/all-targets.json:7]: Plugin SDK code will become part of the given host app code or plugin code, both of which get bundled by webpack.

**Pull request merged**